probe_src/python/probe_py/manual/remote_access.py

```diff
@@ -466,11 +472,66 @@ def lookup_provenance_remote(host: Host, path: pathlib.Path, get_persistent_prov
 
 
 def upload_provenance_local(provenance_info: ProvenanceInfo) -> None:
+    destination_inode_versions, destination_inode_metadatas, augmented_process_closure, augmented_inode_writes = provenance_info
+
+    for inode_version, process_id in augmented_inode_writes.items():
+        inode_version_path = PROCESS_ID_THAT_WROTE_INODE_VERSION / f"{inode_version.str_id()}.json"
+        with inode_version_path.open("w") as f:
+            json.dump(process_id if process_id is not None else None, f)
+
+    for process_id, process in augmented_process_closure.items():
+        process_path = PROCESSES_BY_ID / f"{process_id}.json"
+        with process_path.open("w") as f:
+            json.dump(process, f)
+
     raise NotImplementedError()
 
 
 def upload_provenance_remote(dest: Host, provenance_info: ProvenanceInfo) -> None:
-    raise NotImplementedError()
+    destination_inode_versions, destination_inode_metadatas, augmented_process_closure, augmented_inode_writes = provenance_info
+
+    for inode_version, process_id in augmented_inode_writes.items():
+        if inode_version not in destination_inode_versions:
+            continue
+        inode_version_path = PROCESS_ID_THAT_WROTE_INODE_VERSION / f"{inode_version.str_id()}.json"
+        with inode_version_path.open("w") as f:
+            json.dump(process_id if process_id is not None else None, f)
+    address = dest.get_address()
+    assert address is not None
+    scp_commands = []
+    for inode_version, process_id in augmented_inode_writes.items():
+        inode_version_path = PROCESS_ID_THAT_WROTE_INODE_VERSION / f"{str(inode_version)}.json"
+        scp_commands.append(
+            f"scp {' '.join(dest.scp_options)} {shlex.quote(str(inode_version_path))} {dest.username}@{dest.network_name}:${{process_that_wrote}}/{str(inode_version)}.json"
+        )
+
+    for process_id, process in augmented_process_closure.items():
+        process_path = PROCESSES_BY_ID / f"{process_id}.json"
+        scp_commands.append(
+            f"scp {' '.join(dest.scp_options)} {shlex.quote(str(process_path))} {dest.username}@{dest.network_name}:${{process_by_id}}/{process_id}.json"
+        )
+    combined_scp_command = " && ".join(scp_commands)
+
+    subprocess.run(
+        [
+            "ssh",
+            *dest.ssh_options,
+            address,
+            "sh", "-c", ";".join([
+            "probe_data=${XDG_DATA_HOME:-$HOME/.local/share}/PROBE",
+            "process_that_wrote=${probe_data}/process_id_that_wrote_inode_version",
+            "process_by_id=${probe_data}/processes_by_id",
+            # Ensure the remote directory exists
+            "mkdir -p \"$process_that_wrote\"",
+            "mkdir -p \"$process_by_id\"",
+            combined_scp_command
```

Comment:
I think instead of using `ssh` to call `scp`, we should use the `ssh` to call `echo`, eliminating an extra connection for each file.

The `scp_commands` should be replaced by `echo_commands` that look like this: 

```
f"echo {shlex.quote(json.dumps(insert object here...))} > \"${{process_by_id}}/{id}.json\""
```

A couple of notes:

1. `shlex.quote` helps us quote the string so it doesn't get mangled by shell variables. In particular, it will escape the quotes in JSON with quoted-quotes. I _think_ we only need one `shlex.quote` because it is only being interpreted by one shell (on the remote).
2. Put `$process_by_id` in quotes. It can hypothetically contain a space character (although that would be insane; who would do that?). `process_id` does not need quoting since its an int.

Reply:
- If we establish a single `ssh` connection and then use `scp` to transfer the files, it will not establish a new connection for every transfer. right?
- can we use rsync to transfer multiple files together?

Reply:
Its `ssh` and `echo`, so only 1 ssh connection.

Similar to:

```
ssh server <<EOF
# do commands

echo 'hello world' > path/to/file
EOF
```